File: Minimum Time to Complete Trips.cpp
class Solution {
private:
    bool isPossible(long long mid,vector<int>& time, int totalTrips){
        long long cnt1=0;
        
        for(long long i=0;i<time.size();i++){
            cnt1+=(mid/time[i]);
        }
        
        if(cnt1>=totalTrips) return true;
        else return false;
    }
public:
    long long minimumTime(vector<int>& time, int totalTrips) {
        
        sort(time.begin(),time.end());
        long long sum=0;
        for(auto it:time) sum+=it;

        
        long long lo=1;
        long long hi=sum*totalTrips;
        long long ans=lo;
        
        while(lo<=hi){
            long long mid=lo+(hi-lo)/2;
    
            if(isPossible(mid,time,totalTrips)){
                ans=mid;
                hi=mid-1;
            }
            else lo=mid+1;
        }
        return ans;
    }
};

File: Minimum Time to Complete Trips.java
class Solution {

public long solve(int[] time , long m)
{
    long cal=0;
    for(int val:time)
    {
    
        cal+=(m/val);
    }
    return cal;
}
public long minimumTime(int[] time, long totalTrips) {
    
    
    long sum=0;
    for(int val:time) sum+=val;
        
    
    long l=0;
    long h=sum*totalTrips;
    long ans=h;
    while(l<=h)
    {
        long m=l+(h-l)/2;
       if(solve(time,m)>=totalTrips)
        {
           ans=m;
            h=m-1;
        }
        else
        {
           l=m+1;
        }
    }
    return ans;
}
}
